jsonschema/validators.py

```diff
@@ -56,24 +61,91 @@ def _validates(cls):
     return _validates
 
 
-def create(meta_schema, validators=(), version=None, default_types=None):
+def _generate_legacy_type_checks(types=()):
+    """
+    Generate type check definitions suitable for TypeChecker.redefine_many,
+    using the supplied types. Type Checks are simple isinstance checks,
+    except checking that numbers aren't really bools.
+
+    Arguments:
+
+        types (dict):
+
+            A mapping of type names to their Python Types
+
+    Returns:
+
+        A dictionary of definitions to pass to TypeChecker
+
+    """
+    types = dict(types)
+
+    def gen_type_check(pytypes):
+        pytypes = _utils.flatten(pytypes)
+
+        def type_check(checker, instance):
+            if isinstance(instance, bool):
+                if bool not in pytypes:
+                    return False
+            return isinstance(instance, pytypes)
+
+        return type_check
+
+    definitions = {}
+    for typename, pytypes in iteritems(types):
+        definitions[typename] = gen_type_check(pytypes)
+
+    return definitions
+
+class ValidatorMetaClass(type):
+    @property
+    def DEFAULT_TYPES(self):
+        warn("DEFAULT_TYPES is deprecated, use type_checker",
+             DeprecationWarning)
+        return self._DEFAULT_TYPES
+
+
+def create(meta_schema, validators=(), version=None, default_types=None,
+           type_checker=None):
+
+    use_default_types=False
+    if default_types is not None or type_checker is None:
```

Comment:
This looks like it will mis-warn in the case that someone doesn't provide neither default types nor type checker -- that doesn't seem like it's a deprecated case, is it?

Reply:
I believe so? If neither are provided, we must use the old default behaviour, which relies on default_types? Those that directly use `DraftXValidator` will not see this as these provide a type checker.

Reply:
We have to preserve the old *behavior*, but presumably in a new way -- the old behavior was "if you provide neither thing, you get a validator with type checking via some 'sane' defaults" -- the new way to do that would be to construct a suitably equivalent `TypeChecker`.

But I actually see that the code here is making the "new" default behavior be an empty type checker (with no types registered), but I can't see when that would actually be encountered, which is what's confusing me.

Reply:
> We have to preserve the old behavior, but presumably in a new way -- the old behavior was "if you provide neither thing, you get a validator with type checking via some 'sane' defaults" -- the new way to do that would be to construct a suitably equivalent TypeChecker.

Agreed. 

> But I actually see that the code here is making the "new" default behavior be an empty type checker (with no types registered), but I can't see when that would actually be encountered, which is what's confusing me.

I think maybe the confusion surrounds default _behaviour_ and _default_ values. The type checking behaviour always depends on the value of _both_ `default_types` and `type_checker`. 

The default value for `type_checker` is indeed an empty type checker. The default value for `default_types` is as it used to be, i.e. a sane type mapping. So to maintain the old behaviour - when neither are given, we redefine the empty type checker using the default type mappings. 

Does that make sense? We cannot make the default for `type_checker` anything other than  empty because we need to correctly process non-default-value `default_types`.

Reply:
Is a user who does the following doing something wrong:

```
⊙  ~[jsonschema:python] -c 'import jsonschema; print jsonschema.validators.create({})'                                                                                                                                                                                                                                                                      Julian@Macnetic
jsonschema/validators.py:173: DeprecationWarning: default_types is deprecated, use type_checker
  DeprecationWarning,
<class 'jsonschema.validators.Validator'>
```

they didn't provide `default_types` at all, why do they see a warning about it?

Reply:
My thought process was that they are using it, just implicitly. However, I think you're right and this should not raise a warning. Any future removal of `default_types` should then preserve the same sane type checking defaults through the default value of `type_checker`.